### PR TITLE
perf(BEVFusion): faster bit-identical GPU voxelizer (training)

### PR DESCRIPTION
**Pull request by Max-Bin**
### Summary

This PR adds a fast GPU path for deterministic BEVFusion hard voxelization.

In training, the current deterministic `hard_voxelize` is a major bottleneck, taking around **70 ms per 250k-point LiDAR frame**. The new `voxelize_fast_gpu` keeps deterministic behavior while reducing runtime to about **0.7 ms/frame** on H100, roughly **100× faster**.

The change is backward-compatible and falls back to the existing C++ op on CPU, in non-deterministic mode, or when the voxel count exceeds `max_voxels`.

### Why

The existing deterministic CUDA path is slow mainly because of its algorithm:

- `point_to_voxelidx_kernel` scans previous points to find each point's rank inside its voxel.
- `determin_voxel_num` runs with a single CUDA thread and loops over all points.

For large point clouds, this makes voxelization one of the biggest costs in BEVFusion training.

Using `deterministic=False` is faster, but it is not reproducible because atomic ordering can change which points are kept when a voxel is full. Many training setups therefore keep `deterministic=True` and pay the extra cost.

### What changed

`voxelize_fast_gpu` reproduces the deterministic behavior with parallel PyTorch operations:

1. Filter points by range.
2. Compute voxel coordinates.
3. Stable-sort by voxel id.
4. Keep the original point order inside each voxel.
5. Keep only the first `max_num_points` points per voxel.
6. Scatter the output deterministically.

The voxel row order is different from the C++ op, but this does not affect BEVFusion because downstream sparse encoding uses voxel coordinates. The final BEV output was verified to be identical.

### Verification

Verified against the freshly compiled C++ `hard_voxelize(deterministic=True)` op from this repo on **8 real 120 m LiDAR frames** with around **250k points** and **103k voxels** per frame.

The following matched exactly:

- voxel coordinates
- number of points per voxel
- per-voxel mean features used by `HardSimpleVFE`
- full downstream `extract_feat` output

Result:

```text
max_abs_diff = 0.0
```

### Speed

Benchmark setup: single H100, torch 2.12 + cu130, freshly JIT-compiled op, 5 warm-up runs + 30 measured runs, voxelization only.

```text
C++ deterministic hard_voxelize: 69.8–70.1 ms/frame
voxelize_fast_gpu:              0.69–0.70 ms/frame
speedup:                        ~100×
```

In a downstream E2E trainer, this reduced per-step time from **2.61 s to 0.38 s**, about **6.9× faster**.

### Test

```bash
cd projects/BEVFusion
python -m bevfusion.ops.voxel.test_voxelize_fast
```

The test checks output equivalence and prints benchmark numbers for the current GPU/data.

### CODE REVIEW

File: projects/BEVFusion/bevfusion/ops/voxel/voxelize.py [hunk 1]
```diff
@@ -73,6 +73,66 @@ def forward(ctx, points, voxel_size, coors_range, max_points=35, max_voxels=2000
...
+    """
+    dev = points.device
+    vs = torch.tensor(voxel_size, device=dev, dtype=points.dtype)
+    rmin = torch.tensor(point_cloud_range[:3], device=dev, dtype=points.dtype)
+    rmax = torch.tensor(point_cloud_range[3:], device=dev, dtype=points.dtype)
+    grid = torch.round((rmax - rmin) / vs).long()
+    gx, gy = int(grid[0]), int(grid[1])
+
+    in_range = ((points[:, :3] >= rmin) & (points[:, :3] < rmax)).all(dim=1)
+    pts = points[in_range]
```
[the remark this answers is not in the document]
Comment (Max-Bin) [reply projects/BEVFusion/bevfusion/ops/voxel/voxelize.py]: fixed in 7fc9d22. The fast path now mirrors `dynamic_voxelize_kernel` exactly: it computes the voxel index for every point and keeps only `0 <= coord < grid` (the kernel's `c < 0 || c >= grid` rejection), replacing the raw `>= min & < max` range filter. So for a non-integer `range/voxel_size`, or upper-bound points that floor up to `c == grid`, the same points are now dropped and no out-of-grid voxel is emitted.

Verified bit-identical to the C++ op (coords / num_points / per-voxel mean, `max_diff = 0.0`) on a non-integer grid (`10 / 0.3` → grid 33), exact upper-bound points (`x == 10.0`), and real frames — and the fast output is always within the grid.

**Comment by KSeangTan** [comment]
Thanks @Max-Bin for the implementation. Do you think it is possible to migrate this implementation to autoware.universe as well?

**Comment by Max-Bin** [comment]
> Thanks @Max-Bin for the implementation. Do you think it is possible to migrate this implementation to autoware.universe as well?

@KSeangTan  Since this is a Torch-based method, it is difficult to make it fully compatible with the existing C++ implementation.
Also, the spconv-based deployment on the Autoware side differs from our current results, so further investigation is needed to align them properly.

File: projects/BEVFusion/bevfusion/ops/voxel/voxelize.py [hunk 2]
```diff
@@ -73,6 +73,73 @@ def forward(ctx, points, voxel_size, coors_range, max_points=35, max_voxels=2000
...
+      * ``num_points_per_voxel = min(count, max_num_points)``
+
+    The voxel ROW order differs (unique-sorted here vs. the C++ first-appearance
+    order) but is irrelevant downstream: the sparse encoder indexes voxels by
+    their coords. Returns ``None`` when ``M > max_voxels`` so the caller can fall
+    back to the C++ op, whose voxel-order-dependent truncation we do not
+    replicate in that (rare) case.
+    """
+    dev = points.device
+    vs = torch.tensor(voxel_size, device=dev, dtype=points.dtype)
```
Comment (KSeangTan) [inline projects/BEVFusion/bevfusion/ops/voxel/voxelize.py:108]: Just use
```
device=points.device
```
?

File: projects/BEVFusion/bevfusion/ops/voxel/test_voxelize_fast.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        dim=1,
+    )
+    counts = torch.randint(1, max_pts_per + 1, (num_voxels,), device=dev)
+    rep = cells.repeat_interleave(counts, 0).float()
+    jitter = torch.rand(rep.shape[0], 3, device=dev) * 0.999  # stay inside the cell
+    xyz = (rep + jitter) * vst + rmin
+    return torch.cat([xyz, torch.rand(rep.shape[0], 2, device=dev)], dim=1)
+
+
+def main():
```
Comment (KSeangTan) [inline projects/BEVFusion/bevfusion/ops/voxel/test_voxelize_fast.py]: Should we use unittest framework in python instead?

File: projects/BEVFusion/bevfusion/ops/voxel/voxelize.py [hunk 4]
```diff
@@ -73,6 +73,73 @@ def forward(ctx, points, voxel_size, coors_range, max_points=35, max_voxels=2000
...
+    valid = ((coord >= 0) & (coord < grid)).all(dim=1)
+    pts = points[valid]
+    coord = coord[valid]
+    if pts.shape[0] == 0:
+        return (
+            points.new_zeros((0, max_num_points, feat_dim)),
+            points.new_zeros((0, 3), dtype=torch.int32),
+            points.new_zeros((0,), dtype=torch.int32),
+        )
+    flat = coord[:, 2] * (gy * gx) + coord[:, 1] * gx + coord[:, 0]
```
Comment (KSeangTan) [inline projects/BEVFusion/bevfusion/ops/voxel/voxelize.py:133]: Usually, coord follows the (x, y, z) order, would you mind commenting that do you follow (z, y, x) because of dynamic_voxelize_kernel?

File: projects/BEVFusion/bevfusion/ops/voxel/voxelize.py [hunk 5]
```diff
@@ -73,6 +73,73 @@ def forward(ctx, points, voxel_size, coors_range, max_points=35, max_voxels=2000
...
+
+    order = torch.argsort(flat, stable=True)  # stable -> in-voxel original order
+    flat_s, pts_s, coord_s = flat[order], pts[order], coord[order]
+    uniq, inv, counts = torch.unique_consecutive(flat_s, return_inverse=True, return_counts=True)
+    num_voxels = int(uniq.shape[0])
+    if num_voxels > max_voxels:
```
Comment (KSeangTan) [inline projects/BEVFusion/bevfusion/ops/voxel/voxelize.py:139]: Any reason we can't select the max_voxels here when num_voxels is more?

File: projects/BEVFusion/bevfusion/ops/voxel/voxelize.py [hunk 6]
```diff
@@ -73,6 +73,73 @@ def forward(ctx, points, voxel_size, coors_range, max_points=35, max_voxels=2000
...
+
+    seg_start = torch.zeros(num_voxels, dtype=torch.long, device=dev)
+    seg_start[1:] = torch.cumsum(counts, 0)[:-1]
+    rank = torch.arange(flat_s.shape[0], device=dev) - seg_start[inv]
+    keep = rank < max_num_points
+
+    voxels = torch.zeros(num_voxels, max_num_points, feat_dim, device=dev, dtype=pts.dtype)
+    voxels[inv[keep], rank[keep]] = pts_s[keep]
+    num_points_per_voxel = counts.clamp(max=max_num_points).to(torch.int32)
+    coors = coord_s[seg_start].to(torch.int32)  # (M, 3) = (x, y, z), matching the C++ op
```
Comment (KSeangTan) [inline projects/BEVFusion/bevfusion/ops/voxel/voxelize.py]: Would you mind double checking do they return (x, y, z ) or (z, y, x)?

**Review by KSeangTan — CHANGES_REQUESTED** [review]
LGTM overall, would you mind addressing the comments? Thanks

File: projects/BEVFusion/bevfusion/ops/voxel/voxelize.py [hunk 7]
```diff
@@ -73,6 +73,73 @@ def forward(ctx, points, voxel_size, coors_range, max_points=35, max_voxels=2000
...
+    back to the C++ op, whose voxel-order-dependent truncation we do not
+    replicate in that (rare) case.
+    """
+    dev = points.device
+    vs = torch.tensor(voxel_size, device=dev, dtype=points.dtype)
+    rmin = torch.tensor(point_cloud_range[:3], device=dev, dtype=points.dtype)
+    rmax = torch.tensor(point_cloud_range[3:], device=dev, dtype=points.dtype)
+    grid = torch.round((rmax - rmin) / vs).long()
+    gx, gy = int(grid[0]), int(grid[1])
+
```
[the remark this answers is not in the document]
Comment (Max-Bin) [reply projects/BEVFusion/bevfusion/ops/voxel/voxelize.py]: Fixed. gx/gy/gz are computed on the host from the config now (floor(x+0.5) to match the C++ round()), so no
  more per-forward .item() sync. Still max_abs_diff=0 on the test frames.

File: projects/BEVFusion/bevfusion/ops/voxel/test_voxelize_fast.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    vf, cf, nf = voxelize_fast_gpu(points, voxel_size, pcr, max_points, max_voxels)
+
+    oc, of = torch.argsort(_flat(cc, gx, gy)), torch.argsort(_flat(cf, gx, gy))
+    coord_eq = torch.equal(_flat(cc, gx, gy)[oc], _flat(cf, gx, gy)[of])
+    num_eq = torch.equal(nc[oc], nf[of])
+    # per-voxel mean = HardSimpleVFE input; bit-identical iff kept point-sets match
+    mc = vc.sum(1) / nc.view(-1, 1).float()
+    mf = vf.sum(1) / nf.view(-1, 1).float()
+    vfe_max = (mc[oc] - mf[of]).abs().max().item() if coord_eq else float("nan")
+    ok = coord_eq and num_eq and vfe_max == 0.0
```
[the remark this answers is not in the document]
Comment (Max-Bin) [reply projects/BEVFusion/bevfusion/ops/voxel/test_voxelize_fast.py]: Switched the test to compare the full per-voxel tensors instead of just the means, so it also catches a wrong
  kept-set or slot order. Still 0 diff, including a dense case that hits the max_num_points cap.

File: projects/BEVFusion/bevfusion/ops/voxel/voxelize.py (hunk 2 above)
Comment (Max-Bin) [reply projects/BEVFusion/bevfusion/ops/voxel/voxelize.py:108]: It's a local alias I reuse a bunch of times below, same thing. Can inline it if you prefer.

File: projects/BEVFusion/bevfusion/ops/voxel/test_voxelize_fast.py (hunk 3 above)
Comment (Max-Bin) [reply projects/BEVFusion/bevfusion/ops/voxel/test_voxelize_fast.py]: Done, made it a unittest.TestCase that skips when there's no GPU. Left the benchmark under main.

File: projects/BEVFusion/bevfusion/ops/voxel/voxelize.py (hunk 5 above)
Comment (Max-Bin) [reply projects/BEVFusion/bevfusion/ops/voxel/voxelize.py:139]: It's on purpose. When M > max_voxels the C++ op keeps voxels by first-appearance order, which the sorted path can't reproduce, so truncating here would keep a different set and break the bit-identity. So we fall back to the C++ op instead. It basically never triggers anyway (M ~100k vs max 120k). Added a comment.

File: projects/BEVFusion/bevfusion/ops/voxel/voxelize.py (hunk 6 above)
Comment (Max-Bin) [reply projects/BEVFusion/bevfusion/ops/voxel/voxelize.py]: Checked the kernel: hard_voxelize writes coors as (x, y, z) (coors_offset[0..2] = c_x, c_y, c_z in [voxelization_cuda.cu](https://github.com/tier4/AWML/blob/7a6ec07c851aa21af4c3aee3900eba35c17e03b7/projects/BEVFusion/bevfusion/ops/voxel/src/voxelization_cuda.cu#L56)), and the fast path matches it. The (z,y,x) flip for the sparse encoder is downstream and hits both the same way.

File: projects/BEVFusion/bevfusion/ops/voxel/voxelize.py (hunk 4 above)
Comment (Max-Bin) [reply projects/BEVFusion/bevfusion/ops/voxel/voxelize.py:133]: That's only an internal key to group same-voxel points for the stable sort, it doesn't change the output. coors is still (x, y, z). Added a comment.

File: projects/BEVFusion/bevfusion/ops/voxel/voxelize.py (hunk 2 above)
Comment (KSeangTan) [reply projects/BEVFusion/bevfusion/ops/voxel/voxelize.py:108]: Let's keep it as it is.

**Review by KSeangTan — APPROVED** [review]
The code looks good to me overall. I ran the first test and everything works at the beginning. We can first merge the PR, and then proceeding to test the impact to deployment.
